etl/processor/processor.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2020 Insolar Network Ltd.
+// All rights reserved.
+// This material is licensed under the Insolar License version 1.0,
+// available at https://github.com/insolar/block-explorer/blob/master/LICENSE.md.
+
+package processor
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"sync/atomic"
+
+	"github.com/insolar/block-explorer/etl/interfaces"
+	"github.com/insolar/block-explorer/etl/models"
+	"github.com/insolar/block-explorer/etl/types"
+)
+
+type Processor struct {
+	JDC          <-chan types.JetDrop
+	TaskC        chan Task
+	TaskCCloseMu sync.Mutex
+	Storage      interfaces.Storage
+	Workers      int
+	active       int32
+}
+
+func NewProcessor(jb interfaces.Transformer, storage interfaces.Storage, workers int) *Processor {
+	if workers < 1 {
+		workers = 1
+	}
+	return &Processor{
+		JDC:          jb.GetJetDropsChannel(),
+		Workers:      workers,
+		TaskCCloseMu: sync.Mutex{},
+		Storage:      storage,
+	}
+
+}
+
+var ErrorAlreadyStarted = errors.New("Already started")
+
+func (p *Processor) Start(ctx context.Context) error {
+	p.TaskCCloseMu.Lock()
+	if !atomic.CompareAndSwapInt32(&p.active, 0, 1) {
+		return ErrorAlreadyStarted
+	}
+	p.TaskC = make(chan Task)
+	p.TaskCCloseMu.Unlock()
+
+	for i := 0; i < p.Workers; i++ {
+		go func() {
+			for {
+				t, ok := <-p.TaskC
+				if !ok {
+					return
+				}
+				p.Process(t.JD)
+			}
+		}()
+	}
+
+	go func() {
+		for {
+			jd, ok := <-p.JDC
+			if !ok {
+				p.TaskCCloseMu.Lock()
+				if atomic.CompareAndSwapInt32(&p.active, 1, 0) {
+					close(p.TaskC)
+				}
+				p.TaskCCloseMu.Unlock()
+				return
+			}
+			p.TaskCCloseMu.Lock()
+			if atomic.LoadInt32(&p.active) == 1 {
+				p.TaskC <- Task{&jd}
+			}
+			p.TaskCCloseMu.Unlock()
+		}
+
+	}()
+	return nil
+}
+
+func (p *Processor) Stop(ctx context.Context) error {
+	p.TaskCCloseMu.Lock()
+	if atomic.CompareAndSwapInt32(&p.active, 1, 0) {
+		close(p.TaskC)
+	}
+	p.TaskCCloseMu.Unlock()
+
+	return nil
+}
+
+type Task struct {
```

Comment:
Why do you need it?  I think you can use `types.JetDrop` directly

Reply:
more control in the service